### PR TITLE
[RFC] K_EVENT should not hide the popupmenu

### LINKED ISSUE #4452: Python: accessing vim.current.window closes popupmenu (rplugin)
- Neovim version: NVIM v0.1.3-344-gb030c90
- Operating system: Linux
- Terminal emulator: rxvt-unicode

### Actual behaviour

Accessing `vim.current.window.cursor` in a remote plugin will close any opened popup menu.
This seems to be the case for all properties on `vim.current.window` (at least also for `buffer`), but not `vim.current` itself.  The same applies to calling `vim.funcs.foo()`.

I've noticed this in the context of deoplete, which does this to check if the position has changed:

https://github.com/Shougo/deoplete.nvim/blob/816bea3a9a0c5a1cc53488256d452edea04cb70a/rplugin/python3/deoplete/deoplete.py#L315-L317

It does not seem to be a problem when calling it through `pyeval` in `i_CTRL-R_=`:

> =pyeval('vim.funcs.pumvisible()')

That will insert 1, but keep the popupmenu / completion mode.

This is maybe a bug with https://github.com/neovim/python-client?! /cc @bfredl 

/cc @Shougo

### PR DESCRIPTION
**Pull request by bfredl**
Fixes #4452 (in my own quick testing). Shouldn't be too hard to add a test as well.

Probably `K_FOCUSGAINED`/`LOST` as well...

### CODE REVIEW

**Comment by blueyed** [comment]
Thanks, that fixes the issue for me!

**Comment by justinmk** [comment]
> Probably K_FOCUSGAINED/LOST as well...

Probably. I wonder if there is such a condition for wildmenu and the "Press ENTER" prompt as well, which both are canceled by K_EVENT and FocusGained. https://github.com/neovim/neovim/issues/3416

**Comment by bfredl** [comment]
Hmm, for me "press ENTER" freezes  k_event processing (including "pure" requests like get_current_window that doesn't depend on or change the mode) instead.

**Comment by bfredl** [comment]
Updated with test.

**Comment by blueyed** [comment]
@bfredl 
A probably very related problem is #3757.
Can you check it please?

File: src/nvim/edit.c [hunk 1]
```diff
@@ -3084,8 +3084,10 @@ static bool ins_compl_prep(int c)
 
   /* Ignore end of Select mode mapping and mouse scroll buttons. */
   if (c == K_SELECT || c == K_MOUSEDOWN || c == K_MOUSEUP
-      || c == K_MOUSELEFT || c == K_MOUSERIGHT)
+      || c == K_MOUSELEFT || c == K_MOUSERIGHT || c == K_EVENT
+      || c == K_FOCUSGAINED || c == K_FOCUSLOST) {
     return retval;
```
Comment (justinmk) [inline src/nvim/edit.c]: If this is false, does the K_EVENT key stay on the queue for later processing?

**Comment by justinmk** [comment]
@tarruda @bfredl On second thought, what does it mean to ignore K_EVENT in `ins_compl_prep()`? K_EVENT should not even be _reaching_ `ins_compl_prep()` (correct?). 

Shouldn't K_EVENT keys collect (without processing) on the main queue while user has the completion popup displayed? If we ignore K_EVENT in `ins_compl_prep()` (as in this PR), does that mean a queued message was dropped?

**Comment by bfredl** [comment]
The K_EVENT is still being processed immediately, as evidenced by the test, I don't see what you mean by "queue" here. In completion menu mode (it is not a named mode but it works like a mode) most keys either get a different meaning, or abort the popupmenu and gets processed in "ordinary" insert mode. This if statement defines keys that should just "pass through" (they don't get a different meaning nor do they interrupt completion mode).
Once insert mode gets refactored (e.g, to use a table like normal mode), completion menu could be made a proper submode and this logic much clearer.

**Comment by justinmk** [comment]
Ok, good.

**Comment by Shougo** [comment]
> Once insert mode gets refactored (e.g, to use a table like normal mode), completion menu could be made a proper submode and this logic much clearer.

Yes, Vim should provide completion menu submode like select mode.
I really needed it from some years ago.
It seems be called ctrl-x mode internally.

**Pull request merged**